io-classes-mtl/src/Control/Monad/Class/MonadThrow/Trans.hs

```diff
@@ -0,0 +1,155 @@
+        -> ExceptT e m a -> ExceptT e m a
+      q u (ExceptT b) = ExceptT (u b)
+  uninterruptibleMask f = ExceptT $ uninterruptibleMask $ \u -> runExceptT $ f (q u)
+    where
+      q :: (m (Either e a) -> m (Either e a))
+        -> ExceptT e m a -> ExceptT e m a
+      q u (ExceptT b) = ExceptT (u b)
+
+--
+-- WriterT instances
+--
+
+-- | @since 1.0.0.0
+instance (Monoid w, MonadCatch m) => MonadThrow (WriterT w m) where
+  throwIO = lift . throwIO
+
+-- | @since 1.0.0.0
+instance (Monoid w, MonadCatch m) => MonadCatch (WriterT w m) where
+  catch (WriterT m) f = WriterT $ catch m (runWriterT . f)
+
+  generalBracket acquire release use = WriterT $ fmap f $
+      generalBracket
+        (runWriterT acquire)
+        (\(a, w) e -> g w <$> runWriterT (release a (fst <$> e)))
+        (\(a, w)   -> g w <$> runWriterT (use a))
+    where f ((x,w),(y,z)) = ((x,y),w<>z)
+          g w (a,w') = (a,w<>w')
+
+-- | @since 1.0.0.0
+instance (Monoid w, MonadMask m) => MonadMask (WriterT w m) where
+  mask f = WriterT $ mask $ \u -> runWriterT $ f (q u)
+    where
+      q :: (forall x. m x -> m x)
+        -> WriterT w m a -> WriterT w m a
+      q u (WriterT b) = WriterT (u b)
+  uninterruptibleMask f = WriterT $ uninterruptibleMask $ \u -> runWriterT $ f (q u)
+    where
+      q :: (forall x. m x -> m x)
+        -> WriterT w m a -> WriterT w m a
+      q u (WriterT b) = WriterT (u b)
+
+
+--
+-- RWST Instances
+--
+
+-- | @since 1.0.0.0
+instance (Monoid w, MonadCatch m) => MonadThrow (RWST r w s m) where
+  throwIO = lift . throwIO
+
+-- | @since 1.0.0.0
+instance (Monoid w, MonadCatch m) => MonadCatch (RWST r w s m) where
+  catch (RWST m) f = RWST $ \r s -> catch (m r s) (\e -> runRWST (f e) r s)
+
+  -- | general bracket ignores the state produced by the release callback
+  generalBracket acquire release use = RWST $ \r s -> fmap f $
+      generalBracket
+        (runRWST acquire r s)
+        (\(a, s', w) e -> g w <$> runRWST (release a ((\(x,_,_) -> x) <$> e)) r s')
+        (\(a, s', w)   -> g w <$> runRWST (use a) r s')
+    where f ((x,_,w),(y,s,z)) = ((x,y),s,w<>z)
+          g w (x,s,w') = (x,s,w<>w')
+
+-- | @since 1.0.0.0
+instance (Monoid w, MonadMask m) => MonadMask (RWST r w s m) where
+  mask f = RWST $ \r s -> mask $ \u -> runRWST (f (q u)) r s
+    where
+      q :: (forall x. m x -> m x)
+        -> RWST r w s m a -> RWST r w s m a
+      q u (RWST b) = RWST $ \r s -> u (b r s)
+  uninterruptibleMask f = RWST $ \r s -> uninterruptibleMask $ \u -> runRWST (f (q u)) r s
+    where
+      q :: (forall x. m x -> m x)
+        -> RWST r w s m a -> RWST r w s m a
+      q u (RWST b) = RWST $ \r s -> u (b r s)
+
+--
+-- StateT instances
+--
+
+-- | @since 1.0.0.0
+instance MonadCatch m => MonadThrow (StateT s m) where
+  throwIO = lift . throwIO
+
+-- | @since 1.0.0.0
+instance MonadCatch m => MonadCatch (StateT s m) where
+  catch (StateT m) f = StateT $ \s -> catch (m s) (\e -> runStateT (f e) s)
```

Comment:
This is compatible with what exception package does; it's using [`liftCatch`](https://hackage.haskell.org/package/transformers-0.6.0.4/docs/src/Control.Monad.Trans.State.Strict.html#liftCatch).